### ISSUE TITLE
Error getting zmq4 Go Binding

### ISSUE THREAD

**Issue by marioffmarques** [pebbe/zmq4#45]
Hello,

I'm trying to get zmq4 for use with Go (on windows environment) but i'm getting some errors while getting it with command go get "github.com/pebbe/zmq4". Can anyone provide some kind of information about this? And any tips about install zmq4 go binding on windows ?

![error1](https://cloud.githubusercontent.com/assets/8614153/7295272/fa0f2b92-e968-11e4-925a-ede0a4888e32.jpg)

**Comment by Dipsingh** [comment]
I am running into the same issue on Ubuntu..

**Comment by aletheia7** [comment]
go zmq4 needs the zeromq shared library. Linux needs the /usr/local/lib/libzmq.so and Windows needs a *.dll . The above message indicates the library cannot be found. Either the shared library is not installed, or it's in a location where zmq4 cannot find it.

**Comment by Dipsingh** [comment]
Thanks. I have "/usr/local/lib/libzmq.so" in my linux. I have poked around
few other threads to see what could be wrong but had no luck so far..

**Comment by aletheia7** [comment]
After you  completed `configure`, `make`, `make install`, did you run `ldconfig` as root?

**Comment by Dipsingh** [comment]
Yup.. Followed the below instructions...

git clone https://github.com/zeromq/libzmq
./autogen.sh && ./configure && make -j 4
make check && make install && sudo ldconfig

Just for the sake of trying to see if it makes any difference, I also
tried this installing it in "/usr" as well by supplying the prefix
/usr while running "configure".  Let me know if you need any other
details and I can provide it.

**Comment by aletheia7** [comment]
1. Do you see /usr/local/include/zmq.h?
1. Do you see /usr/local/lib/pkgconfig/libzmq.pc?
1. Are the paths correct in libzmq.pc?

try `git clone https://github.com/zeromq/zeromq4-1`

**Comment by aletheia7** [comment]
Be sure to perform `git checkout -b v4.1.5` after cloning https://github.com/zeromq/zeromq4-1

**Comment by aletheia7** [comment]
Must be root for `make install`

**Comment by Dipsingh** [comment]
1. Do you see /usr/local/include/zmq.h? Yes
   2. Do you see /usr/local/lib/pkgconfig/libzmq.pc? Yes
   3. Are the paths correct in libzmq.pc? seems right ..below is the output

more /usr/local/lib/pkgconfig/libzmq.pc
prefix=/usr/local
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include

Name: libzmq
Description: 0MQ c++ library
Version: 4.1.5
Libs: -L${libdir} -lzmq
Cflags: -I${includedir}

**Comment by Dipsingh** [comment]
Thanks. Let me give it a try..

**Comment by Dipsingh** [comment]
Okay, so here is what I tried. I am not sure what am I missing..

   - git clone https://github.com/zeromq/zeromq4-1
   - cd zeromq4-1/
   - git checkout -b v4.1.5
   - ./autogen.sh
   - ./configure
   - make
   - sudo make install
   - sudo ldconfig


After this I ran the code again "go run zmq.go"
# command-line-arguments
/home/dipsingh/Desktop/GOPROJECT/src/github.com/pebbe/zmq4/errors.go:30:
error: undefined reference to 'zmq_strerror'
/home/dipsingh/Desktop/GOPROJECT/src/github.com/pebbe/zmq4/polling.go:30:
error: undefined reference to 'zmq_poll'
/home/dipsingh/Desktop/GOPROJECT/src/github.com/pebbe/zmq4/polling.go:36:
error: undefined reference to 'zmq_poll'
/home/dipsingh/Desktop/GOPROJECT/src/github.com/pebbe/zmq4/socketget.go:32:
error: undefined reference to 'zmq_getsockopt'
/home/dipsingh/Desktop/GOPROJECT/src/github.com/pebbe/zmq4/socketget.go:38:
error: undefined reference to 'zmq_getsockopt'
/home/dipsingh/Desktop/GOPROJECT/src/github.com/pebbe/zmq4/socketset.go:33:
error: undefined reference to 'zmq_setsockop"
<Output Omitted for brevity>
collect2: error: ld returned 1 exit status

**Comment by aletheia7** [comment]
Do you have pkg-config installed? `apt-get install pkg-config`

**Comment by Dipsingh** [comment]
Yeah i have pkg-config installed.

pkg-config --version
0.26

**Comment by aletheia7** [comment]
Do you have the Ubuntu libzmq* package installed?

```
dpkg-query -l libzmq\*
```

**Comment by Dipsingh** [comment]
This is what i have.

dpkg-query -l libzmq\*
Desired=Unknown/Install/Remove/Purge/Hold
|
Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                                 Version
Architecture            Description
+++-====================================-=======================-=======================-==============================================================================
ii  libzmq-dev:amd64                     2.2.0+dfsg-5            amd64
              lightweight messaging kernel (development files)
ii  libzmq1:amd64                        2.2.0+dfsg-5            amd64
              lightweight messaging kernel (shared library)

**Comment by aletheia7** [comment]
Those are old libraries. Do you want to remove them? If you remove them, other things might also be removed. What version of Ubuntu? `lsb_release -a`

`apt-get remove libzmq-dev:amd64 libzmq:amd64`

**Comment by Dipsingh** [comment]
Yeah i can remove them, which ones i need to look for ? ..its a test box

lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:   Ubuntu 14.04.3 LTS
Release:       14.04
Codename:       trusty

**Comment by aletheia7** [comment]
Run the apt-get remove above. Run ldconfig. Try configure, make, make install, ldconfig again.

**Comment by Dipsingh** [comment]
>Try configure, make, make install, ldconfig again
just to clarify, run above commands on zeromq ?

**Comment by aletheia7** [comment]
Yes.

**Comment by Dipsingh** [comment]
removed the package, ran  "sudo ldconfig". Then I ran the configure,make,
sudo make install and sudo ldconfig but no luck ..ran into the same error
..Thanks for helping me so far..

dpkg-query -l libzmq\*
dpkg-query: no packages found matching libzmq*

**Comment by aletheia7** [comment]
You're welcome.
What is the output of `type pkg-config`

**Comment by Dipsingh** [comment]
Here is the output

type pkg-config
pkg-config is /usr/bin/pkg-config

**Comment by aletheia7** [comment]
What is the output of `go env`

**Comment by Dipsingh** [comment]
go env
GOARCH="amd64"
GOBIN=""
GOCHAR="6"
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/dipsingh/Desktop/GOPROJECT"
GORACE=""
GOROOT="/usr"
GOTOOLDIR="/usr/lib/gccgo/tool"
TERM="dumb"
CC="gcc"
GOGCCFLAGS="-g -O2 -fPIC -m64 -pthread"
CXX="g++"
CGO_ENABLED="1"

**Comment by aletheia7** [comment]
1. Did you compile go from source?
1. What is the output from `dpkg-query -l go\*`

**Comment by aletheia7** [comment]
Go is normally installed under /usr/lib/go.

**Comment by Dipsingh** [comment]
1. Did you compile go from source? Honestly don't recall, but I
   don't think I compiled it..
   2. What is the output from dpkg-query -l go\*

dpkg-query -l go\*
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait
/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                                 Version
Architecture            Description
+++-====================================-=======================-=======================-==============================================================================
un  go-compiler                          <none>                  <none>
             (no description available)
un  gobject-introspection-freedesktop    <none>                  <none>
             (no description available)
un  gobject-introspection-glib-2.0       <none>                  <none>
             (no description available)
un  gobject-introspection-repository     <none>                  <none>
             (no description available)
un  golang-go                            <none>                  <none>
             (no description available)

**Comment by Dipsingh** [comment]
Alright ..Now I can bang my head in the wall..I just pulled golang packages
again and it worked ...

Gezzz..

**Comment by aletheia7** [comment]
Great. What is your go version: `go version`

Your version of Ubuntu is getting quite old. You should upgrade.

I ran Ubuntu of years. Ubuntu's upstream is Debian. I prefer Debian Stretch/Testing with XFCE Desktop over Ubuntu.

**Comment by Dipsingh** [comment]
I have 1.6..

Yeah, I know the Ubuntu version is old, but I had to keep it for some app
dependency.. I agree that I need to look for other options..

Thanks for the pointers and all the help.. Really appreciate it..

**Comment by aletheia7** [comment]
Your welcome.
